plugin/adsk/scripts/mayaUsdTranslatorExport.mel

```diff
@@ -41,16 +108,134 @@ global proc int mayaUsdTranslatorExport (string $parent,
+                    intFieldGrp -e -v2 $endTime framRangeFields;
+                } else if ($optionBreakDown[0] == "frameStride") {
+                    float $frameStride = $optionBreakDown[1];
+                    floatFieldGrp -e -v1 $frameStride frameStrideField;
+                } else if ($optionBreakDown[0] == "parentScope") {
+                    mayaUsdTranslatorExport_SetTextField($optionBreakDown[1], "parentScopeField");
+                } else if ($optionBreakDown[0] == "exportInstances") {
+                    mayaUsdTranslatorExport_SetCheckbox($optionBreakDown[1], "exportInstancesCheckBox");                    
+                } else if ($optionBreakDown[0] == "exportVisibility") {
+                    mayaUsdTranslatorExport_SetCheckbox($optionBreakDown[1], "exportVisibilityCheckBox");
+                } else if ($optionBreakDown[0] == "mergeTransformAndShape") {
+                    mayaUsdTranslatorExport_SetCheckbox($optionBreakDown[1], "mergeTransformAndShapeCheckBox");                    
+                } else if ($optionBreakDown[0] == "stripNamespaces") {
+                    mayaUsdTranslatorExport_SetCheckbox($optionBreakDown[1], "stripNamespacesCheckBox");                    
+                }
+            }
+        }
+        // Set visibility for anim widgets
+        mayaUsdTranslatorExport_AnimationCB();
+
+        $bResult = 1;
+
     } else if ($action == "query") {
-        string $currentOptions = "";
+        $currentOptions = mayaUsdTranslatorExport_AppendFromCheckbox($currentOptions, "exportUVs", "exportUVsCheckBox");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromCheckbox($currentOptions, "exportColorSets", "exportColorSetsCheckBox");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromPopup($currentOptions, "defaultMeshScheme", "defaultMeshSchemePopup");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromCheckbox($currentOptions, "animation", "animationCheckBox");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromCheckbox($currentOptions, "eulerFilter", "eulerFilterCheckBox");
+        $currentOptions = mayaUsdTranslatorExport_AppendFrameRange($currentOptions);
+        $currentOptions = mayaUsdTranslatorExport_AppendFromFloatField($currentOptions, "frameStride", "frameStrideField");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromTextField($currentOptions, "parentScope", "parentScopeField");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromCheckbox($currentOptions, "exportInstances", "exportInstancesCheckBox");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromCheckbox($currentOptions, "exportVisibility", "exportVisibilityCheckBox");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromCheckbox($currentOptions, "mergeTransformAndShape", "mergeTransformAndShapeCheckBox");
+        $currentOptions = mayaUsdTranslatorExport_AppendFromCheckbox($currentOptions, "stripNamespaces", "stripNamespacesCheckBox");
```

Comment:
FWIW, we've banned people writing boiler plate MEL code here. It takes too much dev effort, it's bug prone, and it's usually trivial to auto generate the code. For reference, this is the method that generates our import/export dialogs: 

https://github.com/Autodesk/maya-usd/blob/dev/plugin/al/mayautils/AL/maya/utils/PluginTranslatorOptions.cpp#L712

And here's an example of how we're specifying the options:
https://github.com/Autodesk/maya-usd/blob/dev/plugin/al/lib/AL_USDMaya/AL/usdmaya/fileio/ExportTranslator.h#L74 

All of our AE templates, menus, command dialogs, export dialogs are all autogenerated. The code generators are all within our mayautils lib here:
https://github.com/Autodesk/maya-usd/tree/dev/plugin/al/mayautils/AL/maya/utils

Reply:
Boilerplate is when you write/copy the exact same chunk of code over and over again.  Extracting a chunk of code into its own function and then calling that function in multiple places is the opposite of “boilerplate code” is it not?  That is what’s going on here, each of the checkbox’s or float fields call the exact same function with their option/control names passed in.

I’ve actually had to debug a bit of the C++ auto-generated MEL in the AL plugin and I’ll admit that I wasn’t a fan of that experience.  I’m not the original author of most of this so I don’t have a really strong opinion about a lot of it (almost everything here comes from the Pixar plugin), but I’m definitely not anti-MEL and wouldn’t want to move all of this into compiled C++ for no real benefit.  E.g. everything in PluginTranslatorOptions.cpp is actually MEL…a developer had to write all of that in MEL, debug all of that in MEL, before copying it into C++, adding a ton of casts, making sure they got all of their strings quoted correctly, before compiling it.  I’m not sure I’d agree with that approach being easier and less bug prone.  I’m sure that once it’s done it’s pretty much never going to change again, but I think it’s the same for the functions here.  

The one thing I would like to change is to move some of this into a shared file because I think there are exact copies of all of this in the import .mel file.


Now I have to say Rob that it’s quite ironic that you and I are talking about MEL today because I just spent an hour looking at your web page.  A Maya user had hacked together a script based on some snippets they grabbed from your MEL tutorials and couldn’t figure out how to make it all work so they contacted our Support for help.  At least I’m assuming it’s your page…I can’t imagine there being more than one Maya expert named RobTheBloke 😊